### ISSUE TITLE
Populate a map of OID -> pkix.Extension in Certificate objects.

### ISSUE THREAD

**Issue by cpu** [zmap/zcrypto#198]
**tl;dr - I think `zcrypto` should offer a `map[string]pkix.Extension` field on `x509.Certificate` to avoid lots of linear searching in `zlint`'s many lints.**

In [zlint](https://gitub.com/zmap/zlint) a common pattern in the `CheckApplies(c *x509.Certificate) bool` functions of individual lints is to check a `zcrypto.X509.Certificate` for a specific x509 extension. e.g. consider the [`lint_ext_san_empty_name`](https://github.com/zmap/zlint/blob/d9a29c3ddfb5b3b498dea53607b9c395267d3807/lints/lint_ext_san_empty_name.go#L42-L44) lint's `CheckApplies` function:

```
return util.IsExtInCert(c, util.SubjectAlternateNameOID)
```

The `util.IsExtInCert` function in turn relies on a `util.GetExtFromCert` function that [performs a linear search](https://github.com/zmap/zlint/blob/d9a29c3ddfb5b3b498dea53607b9c395267d3807/util/oid.go#L110-L119) over the `zcrypto.X509.Certificate`'s `Extensions` field:
```
	for i := range cert.Extensions {
		if oid.Equal(cert.Extensions[i].Id) {
			return &(cert.Extensions[i])
		}
	}
```

Looking across the whole `zlint` codebase with a quick `ack` I see there are **200** references to `util.GetExtFromCert` or `util.IsExtInCert`:

```
$ ack -c -h "GetExtFromCert|IsExtInCert" $GOPATH/src/github.com/zmap/zlint
200
```

Benchmarking confirms that ` github.com/zmap/zlint/util.GetExtFromCert` and `encoding/asn1.ObjectIdentifier.Equal` are present int the top 10 nodes accounting for cpu time in a `pprof` trace:

```
$ cd $GOPATH/src/github.com/zmap/zlint
$ go test --run=XXX -bench=. -cpuprofile all.profile
<snipped>
$ go tool pprof all.profile
File: zlint.test
Type: cpu
Time: Jul 26, 2019 at 10:58am (EDT)
Duration: 4.13mins, Total samples = 4.79mins (115.90%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top10
Showing nodes accounting for 146.71s, 51.07% of 287.25s total
Dropped 522 nodes (cum <= 1.44s)
Showing top 10 nodes out of 221
      flat  flat%   sum%        cum   cum%
    45.89s 15.98% 15.98%    142.36s 49.56%  runtime.mallocgc
    31.59s 11.00% 26.97%     39.26s 13.67%  runtime.heapBitsSetType
    18.20s  6.34% 33.31%     18.20s  6.34%  runtime.nextFreeFast
    11.27s  3.92% 37.23%     20.05s  6.98%  runtime.scanobject
     9.90s  3.45% 40.68%      9.90s  3.45%  runtime.memclrNoHeapPointers
     8.31s  2.89% 43.57%      8.31s  2.89%  encoding/asn1.ObjectIdentifier.Equal
     6.92s  2.41% 45.98%     13.52s  4.71%  github.com/zmap/zlint/util.GetExtFromCert
     5.63s  1.96% 47.94%         7s  2.44%  runtime.heapBitsForAddr
     4.60s  1.60% 49.54%    130.51s 45.43%  runtime.newobject
     4.40s  1.53% 51.07%    245.67s 85.52%  github.com/zmap/zlint.BenchmarkZlint.func4
```

The `CheckApplies` function is called for all configured lints, for all input certificates because the function is used to gate the further application of the lint logic. 

It seems like much of this processing could be avoided if `zcrypto` were to provide a new field on `x509.Certificate` that holds `map[string]pkix.Extension` where the `string` key is the `asn1.ObjectIdentifier.String()` of the `pkix.Extension.Id`.

This map could be populated once at the time the certificate was parsed. Subsequently all of the lints that need to check if a given certificate have an extension they are interested in can do it with an O(1) membership check on the map field.

What do folks think? Am I missing an obvious downside (other than slightly more memory per Certificate)?  If you're game I will take a crack at implementing this and then adjusting the downstream lints afterwards.

cc @dadrian @zakird

**Comment by zakird** [comment]
Sounds very reasonable to me.

**Comment by cpu** [comment]
Cool, I think I can put up a PR against `zcrypto` early next week.

**Comment by cpu** [comment]
> Cool, I think I can put up a PR against zcrypto early next week.

Here's a PR for review: https://github.com/zmap/zcrypto/pull/200

PTAL, thanks!
